src/main/java/org/embulk/input/marketo/delegate/ProgramInputPlugin.java

```diff
@@ -0,0 +1,224 @@
+
+            if (earliest.isAfter(DateTime.now())) {
+                throw new ConfigException(String.format("`earliest_updated_at` (%s) cannot precede the current date (%s)",
+                                earliest.toString(DATE_FORMATER),
+                                (DateTime.now().toString(DATE_FORMATER))));
+            }
+
+            if (earliest.isAfter(latest)) {
+                throw new ConfigException(String.format("Invalid date range. `earliest_updated_at` (%s) cannot precede the `latest_updated_at` (%s).",
+                                earliest.toString(DATE_FORMATER),
+                                latest.toString(DATE_FORMATER)));
+            }
+            // if filter type is selected, filter value must be presented
+            if (task.getFilterType().isPresent() && (!task.getFilterValues().isPresent() || task.getFilterValues().get().isEmpty())) {
+                throw new ConfigException("filter_value is required when selected filter_type");
+            }
+        }
+    }
+
+    @Override
+    protected Iterator<ServiceRecord> getServiceRecords(MarketoService marketoService, PluginTask task)
+    {
+        Iterable<ObjectNode> nodes = null;
+        if (!task.getQueryBy().isPresent()) {
+            nodes = marketoService.getPrograms();
+        }
+        else if (task.getQueryBy().isPresent() && task.getQueryBy().get() == QueryBy.TAG_TYPE) {
+            nodes = marketoService.getProgramsByTag(task.getTagType().get(), task.getTagVallue().get());
+        }
+        else if (task.getQueryBy().isPresent() && task.getQueryBy().get() == QueryBy.DATE_RANGE) {
+            nodes = marketoService.getProgramsByDateRange(task.getEarliestUpdatedAt().get(),
+                            task.getLatestUpdatedAt().get(),
+                            task.getFilterType().orNull(),
+                            task.getFilterValues().orNull());
+        }
+
+        return FluentIterable.from(nodes).transform(MarketoUtils.TRANSFORM_OBJECT_TO_JACKSON_SERVICE_RECORD_FUNCTION).iterator();
+    }
+
+    @Override
+    public ConfigDiff buildConfigDiff(PluginTask task, Schema schema, int taskCount, List<TaskReport> taskReports)
+    {
+        ConfigDiff configDiff = super.buildConfigDiff(task, schema, taskCount, taskReports);
+        // set next next earliestUpdatedAt, latestUpdatedAt
+        if (task.getQueryBy().isPresent() && task.getQueryBy().get() == QueryBy.DATE_RANGE && task.getIncremental()) {
+
+            DateTime earliest = new DateTime(task.getEarliestUpdatedAt().get());
+            DateTime latest = new DateTime(task.getLatestUpdatedAt().get());
+
+            Duration d = new Duration(earliest, latest);
+            DateTime nextEarliestUpdatedAt = latest.plusSeconds(1);
+
+            configDiff.set("earliest_updated_at", nextEarliestUpdatedAt.toString(DATE_FORMATER));
+            if (task.getReportDuration().isPresent()) {
+                configDiff.set("report_duration", task.getReportDuration());
```

Comment:
`Optional`  method has `or` function will be executed if the optional value is not presented: https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Optional.html#or(T), should use that one instead of `if` `else`